### PR TITLE
Updating Vector*.IsHardwareAccelerated to be recursive

### LINKED ISSUE #38162: Vector.IsHardwareAccelerated result differs when called directly and called via reflection
```C#
using System;
using System.Reflection;
using System.Numerics;

class X
{
    public static int Main()
    {
        bool hw1 = Vector.IsHardwareAccelerated;

        Assembly spc = typeof(object).Assembly;
        Type vecType = spc.GetType("System.Numerics.Vector");
        MethodInfo method = vecType.GetMethod("get_IsHardwareAccelerated");
        bool hw2 = (bool) method.Invoke(null, null);

        Console.WriteLine($"Direct call to Vector.IsHardwareAccelerated returns {hw1}");
        Console.WriteLine($"Reflection call to Vector.IsHardwareAccelerated returns {hw2}");

        return (hw1 == hw2 ? 100 : -1);
    }
}
```
results in
```
Direct call to Vector.IsHardwareAccelerated returns True
Reflection call to Vector.IsHardwareAccelerated returns False
```
Seems like we're not recognizing this as an intrinsic when it is jitted on its own.

category:cq
theme:vector-codegen
skill-level:intermediate
cost:medium

### PR DESCRIPTION
**Pull request by tannergooding**
This resolves https://github.com/dotnet/runtime/issues/38162

This updates the various `IsHardwareAccelerated` APIs to be recursive, just as they are for the `IsSupported` APIs. This ensures that they return the same result whether invoked directly or indirectly (such as via reflection). This ensures that the debugger and other scenarios report a consistent worldview.

### CODE REVIEW

File: src/libraries/System.Runtime.Intrinsics/tests/Vectors/Vector128Tests.cs [hunk 1]
```diff
@@ -8,6 +8,13 @@ namespace System.Runtime.Intrinsics.Tests.Vectors
 {
     public sealed class Vector128Tests
     {
+        [Fact]
+        public unsafe void Vector128IsHardwareAcceleratedTest()
+        {
+            var methodInfo = typeof(Vector128).GetMethod("get_IsHardwareAccelerated");
```
Comment (stephentoub) [inline src/libraries/System.Runtime.Intrinsics/tests/Vectors/Vector128Tests.cs:14]: nit
```suggestion
            MethodInfo methodInfo = typeof(Vector128).GetMethod("get_IsHardwareAccelerated");
```

File: src/libraries/System.Runtime.Intrinsics/tests/Vectors/Vector256Tests.cs [hunk 2]
```diff
@@ -8,6 +8,13 @@ namespace System.Runtime.Intrinsics.Tests.Vectors
 {
     public sealed class Vector256Tests
     {
+        [Fact]
+        public unsafe void Vector256IsHardwareAcceleratedTest()
+        {
+            var methodInfo = typeof(Vector256).GetMethod("get_IsHardwareAccelerated");
```
Comment (stephentoub) [inline src/libraries/System.Runtime.Intrinsics/tests/Vectors/Vector256Tests.cs:14]: nit
```suggestion
            MethodInfo methodInfo = typeof(Vector256).GetMethod("get_IsHardwareAccelerated");
```

File: src/libraries/System.Runtime.Intrinsics/tests/Vectors/Vector64Tests.cs [hunk 3]
```diff
@@ -8,6 +8,13 @@ namespace System.Runtime.Intrinsics.Tests.Vectors
 {
     public sealed class Vector64Tests
     {
+        [Fact]
+        public unsafe void Vector64IsHardwareAcceleratedTest()
+        {
+            var methodInfo = typeof(Vector64).GetMethod("get_IsHardwareAccelerated");
```
Comment (stephentoub) [inline src/libraries/System.Runtime.Intrinsics/tests/Vectors/Vector64Tests.cs:14]: nit
```suggestion
            MethodInfo methodInfo = typeof(Vector64).GetMethod("get_IsHardwareAccelerated");
```

**Review by stephentoub — APPROVED** [review]

**Comment by tannergooding** [comment]
@davidwrighton, looks like `ReadyToRun` for `Vector256.IsHardwareAccelerated` is returning `false` for the indirect invocation (meaning its baked into `corelib` as `false`).

Is that expected? We don't appear to do the same for `Vector<T>` or `Avx/Avx2`

File: src/libraries/System.Numerics.Vectors/tests/GenericVectorTests.cs [hunk 4]
```diff
@@ -25,6 +25,13 @@ static GenericVectorTests()
             dummy = System.Numerics.Vector<float>.One;
         }
 
+        [Fact]
+        public unsafe void IsHardwareAcceleratedTest()
+        {
+            var methodInfo = typeof(Vector).GetMethod("get_IsHardwareAccelerated");
```
Comment (tannergooding) [inline src/libraries/System.Numerics.Vectors/tests/GenericVectorTests.cs:31]:
```suggestion
            MethodInfo methodInfo = typeof(Vector).GetMethod("get_IsHardwareAccelerated");
```

**Comment by tannergooding** [comment]
CC. @dotnet/crossgen-contrib for help with the unexpected failure above.

**Comment by jeffhandley** [comment]
@davidwrighton Do you think there's something lurking in these unexpected failures that we need to make sure to address in 7.0?

**Comment by danmoseley** [comment]
I think I reported this originally - we're trying to make writing vectorized code more accessible in .NET 7, and this is likely to confuse initially as it did me, so it would be nice to get it in .NET 7.

**Comment by stephentoub** [comment]
@davidwrighton, are you still able to look at this?

**Comment by danmoseley** [comment]
@davidwrighton ?

**Comment by davidwrighton** [comment]
Sorry, this fell off my radar. I'm looking at it today.

**Comment by tannergooding** [comment]
@fanyang-mono, could you assist with the changes needed for Mono here?

I went and looked at `intrinsics.c` and `simd-intrinsics.c`, but it looks like the `get_IsHardwareAccelerated` methods already have handling and similar handling to `get_IsSupported`, so its not clear what's missing and causing it to fail on the recursion rather than always creating a constant true/false node

**Comment by fanyang-mono** [comment]
@tannergooding I will see if I have time to get to it today or tmr. Otherwise, I will forward this to another person on mono team to handle. Because I will be on vacation starting from this Friday for two weeks.

**Comment by tannergooding** [comment]
Thanks much!

**Comment by danmoseley** [comment]
@fanyang-mono did you manage to identify someone who could help with the mono side?

**Comment by fanyang-mono** [comment]
@vargaz Could you help out with this while I am on vacation?

**Comment by tannergooding** [comment]
ping @fanyang-mono, @vargaz on this.

I imagine its too late for .NET 7 at this point, but it would still be nice to get fixed

**Comment by danmoseley** [comment]
@tannergooding is the CoreCLR change mergeable on its own?

**Comment by danmoseley** [comment]
Also, I wouldn't rule out .NET 7 if the risk is low, as it's a goal of the release to make it easy to use vectorization in your own code. This trips you up right away.

**Comment by tannergooding** [comment]
> @tannergooding is the CoreCLR change mergeable on its own?

Technically, yes. But there's no real point, IMO, if we aren't going to ship Corelib with the support since users will never trigger the code.

> Also, I wouldn't rule out .NET 7 if the risk is low, as it's a goal of the release to make it easy to use vectorization in your own code. This trips you up right away.

I'm not sure I'd say this is a right away scenario. Calling this via reflection or indirectly such as via a delegate is expected to be an extremely rare/niche scenario and not the normal usage pattern.

I expect the debugger case is more likely, but we really need the Mono support as well.

**Comment by danmoseley** [comment]
Ah. Agreed, I think the debugger case is the only one that is significant.

**Comment by danmoseley** [comment]
cc @SamMonoRT since it's potentially 7.0

**Comment by vargaz** [comment]
This will fix the mono interpreter failures:
https://github.com/vargaz/runtime/commit/8be6b8dbb2968d416924c49759aafcf89241f81b
Were there any other failures ?

**Comment by danmoseley** [comment]
relevant failures eg
```
Assert failure(PID 7492 [0x00001d44], Thread: 5828 [0x16c4]): Assertion failed '!"Unhandled must expand intrinsic, throwing PlatformNotSupportedException"' in 'System.Runtime.Intrinsics.Vector256:get_IsHardwareAccelerated():bool' during 'Importation' (IL size 6; hash 0x068ee1df; FullOpts)

File: D:\a\_work\1\s\src\coreclr\jit\importer.cpp Line: 4938
Image: C:\h\w\A5E108DA\p\corerun.exe
```
and mono failures eg JIT.HardwareIntrinsics.X86.Bmi1

**Comment by tannergooding** [comment]
Its actually not clear to me why CI is asserting. It doesn't assert locally at all.

However, I've added an extra safety path to see if that resolves the issue.

**Comment by tannergooding** [comment]
Actually, realized the potential issue right as I typed that up. The path that could convert `IsHardwareAccelerated` to `true` was outside the `FEATURE_HW_INTRINSICS` path when it should've been "in" with the fallback handler always returning `false`.

**Comment by tannergooding** [comment]
> Were there any other failures ?

@vargaz, looks like still some Mono Linux failures where its unexpectedly getting `System.PlatformNotSupportedException`.

-----

There is also a failure for RyuJIT `x64/x86 Checked no_tiered_compilation`. It's saying:
```
Assert failure(PID 7696 [0x00001e10], Thread: 5688 [0x1638]): Assertion failed '!"Unhandled must expand intrinsic, throwing PlatformNotSupportedException"' in 'System.Runtime.Intrinsics.Vector128:get_IsHardwareAccelerated():bool' during 'Importation' (IL size 6; hash 0x330ddaf5; FullOpts)

File: D:\a\_work\1\s\src\coreclr\jit\importer.cpp Line: 4938
```

@davidwrighton, this appears to be because these tests disable SSE2 and so the intrinsic returned is `NI_IsSupported_Dynamic`. I don't see any handling for this in the JIT, so what's supposed to happen for it in this recursive case?

**Comment by vargaz** [comment]
The mono failures are also present in HEAD, so they are unrelated:
https://github.com/dotnet/runtime/runs/8369467985

**Comment by tannergooding** [comment]
Failures are https://github.com/dotnet/runtime/issues/75753 and https://github.com/dotnet/runtime/issues/75606

**Comment by tannergooding** [comment]
@danmoseley, are you still wanting this backported to .NET 7?

For `Vector64/128/256<T>` this is "new API functionality". For `Vector<T>` this has been the behavior since it was released in 2014.

There have been only a couple user reports on this, so I'm not sure if it meets the bar or not.

**Comment by danmoseley** [comment]
Yeah now we branched, I don't think it does. Thanks for driving it though.

**Pull request merged**